### PR TITLE
Topic recognition: 47434673 StyleGAN2 Brain Image Generator

### PR DESCRIPTION
**Pull request by ThomBarros**
Hello COMP3710 team,

I am student s4743467 (Thomas Barros) and I would like to merge my solution to the topic recognition problem to the topic-recognition branch of PatternAnalysis-2024. My solution is a StyleGAN2 model trained on the OASIS dataset. The files I have added are contained under the recognition subdirectory under the directory 47434673_StyleGAN2. The files added are:

- `config.py`: Contains all the hyperparameters. Also contains the path to the data directory.
- `dataset.py`: Contains the `get_data` method which preproccesses the data and creates the dataloader. Also contains the `show_imgs` method which displays some of the training images.
- `utils.py`: Contains methods for data augmentation and optimisation.
- `modules.py`: Contains architecture of the mapping network, discriminator and generator.
- `predict.py`: Contains `plot_loss` method for plotting the discriminator and generator loss as well as `generate_examples` method for generating some images using the trained model.
- `train.py`: The main file of the project. Contains the functionality for training the models, as well as execute all other functionality of the project. This is the file that is called to run the project.

Note also that an `assets` folder has been included, which contains the images used in the README/report.

Please refer to the README file for the report as well as instructions on how to navigate the repository. 

Please let me know of any changes that need to be made.

Thomas Barros 47434673

### CODE REVIEW

**Comment by ThomBarros** [comment]
I thought I might just add that I originally started this project on Google colab, which is why you might see that the first few commits are just massive file additions, as I was moving all of my work from colab onto python files.

**Comment by ThomBarros** [comment]
Hi, so I just figured out that I in fact used the OASIS dataset to train my StyleGAN2 and not the ADNI dataset. Originally I thought we were using the OASIS dataset (same as Demo 2), but I just reread the task sheet again to make sure I had completed everything for the final submission, and then I realised my mistake. As a result, you'll notice that my UMAP was a kind of cobbled together solution, as the OASIS dataset doesn't have any ground truth labels, so I had to improvise (see the README for details).

**Comment by Uto0op** [comment]
Recognition problem: StyleGAN on ADNI ---- Hard Difficulty

1. Algorithm solves the problem: (2.5 Marks)
Somewhat reasonably clear generated images;
The problem is on ADNI data, not OASIS -2.5

2. Algorithm demonstrated and functions: (3 Marks)

4. Design: (1 Mark)
Modularized design and good practice using utils script

5. Commenting:
Sufficient docstrings and comments (1 Mark)

6. Algorithm above Normal Difficulty: (5 Marks)

7. Algorithm is Hard difficulty: (5 Marks)

**Suggestions**:
- The most noticeable issue, as you mentioned, is the model was trained/used on a wrong dataset which is not what this problem asks for, so I'm taking half the marks off from the "problem solving" perspective. You can incorporate by simply retraining the model on ADNI and reproducing the results shown on the README.md;
- `config.py` and `colour_umap.py` are not needed for this project, you could merge them into `utils.py`;
- It's good to monitor progressive training outputs, but you might also want to justify the loss plots provided.

**Comment by ThomBarros** [comment]
Thank you for your feedback, I will make changes to the project and report based on your suggestions.

**Comment by yexincheng** [comment]
**This is an initial inspection, no action is required at this point**
1. **Solves problem**: Nince visualization. A good introduction to networks but could have more detail. Dependency versions were not specified. Part of the training log was shown in a screenshot, the training loss plots of different epochs were provided. The generated images from the intermediate checkpoints were provided. The final generated image is reasonable. UMAP embedding plot is provided. Image size should be included in the report and readme. There were two images smaller than the others; do they have the same size?
2. **Implementation**: Clear and well structured
3. **Design**: Good
4. **Comment**: Basic comments, could have more detail like input and output.
5. **Difficulty**: Hard

**Note**
The student used the wrong dataset OASIS rather than ADNI

**Comment by ThomBarros** [comment]
Understood, I will not implement any changes to the existing project until I receive the explicit instructions to. Thank you for this clarification.

**Comment by gayanku** [comment]
## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>Incorrect design and implementation. The problem is on ADNI data, not OASIS.</td><td>-2</td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>Header blocks. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>Incorrect solution to problem + low image quality.</td><td>-2</td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. </td><td></td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting present. </td><td></td></tr>
	<tr><td></td><td>No Data leakage found. </td><td></td></tr>
	<tr><td></td><td>Difficulty : Hard. StyleGAN (Hard Difficulty)</td><td></td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Good Meaningful commit messages. </td><td></td></tr>
	<tr><td></td><td>Some/Adequate Progressive commits. Somewhat limited.</td><td>-1</td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Acceptable. Some areas in the report is hard to comprehend.</td><td>-1</td></tr>
	<tr><td></td><td>Model/technical explanation :Good. </td><td></td></tr>
	<tr><td></td><td>Description and Comments :Good. </td><td></td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. PDF Checked.</td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch). ON time.</td><td></td></tr>
	<tr><td></td><td>No Feedback required. </td><td></td></tr>
	<tr><td></td><td>Request Description is good. </td><td></td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>-6</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_

**Comment by ThomBarros** [comment]
Hello team, thank you for the feedback. I was just wondering if there is anything I need to fix based on the feedback? In particular, I'm kicking myself for using the wrong dataset, it will probably make me go from a 7 to a 6 in the course. Is there any possibility of me rectifying this by making adjustments to the project? Thanks.

**Comment by shakes76** [comment]
Unfortunately no as we can only mark the PR's main content (non-merge related aspects) at the due date for fairness.
